File: ArchiveCapture/app/src/main/java/com/archiveprocessor/capture/net/MacClient.kt
package com.archiveprocessor.capture.net

import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.util.concurrent.TimeUnit

/** Why a pairing attempt failed — so the UI can name the cause + the fix instead of a bare "couldn't
 *  reach". UNREACHABLE = timeout / no route (the client-isolation case); REFUSED = reached the network
 *  but nothing listening (server not started / wrong port); UNAUTHORIZED = reached the Mac, bad token. */
enum class Reachability { OK, UNAUTHORIZED, REFUSED, UNREACHABLE }

/** Protocol v2 client for the Archive Processor Live Capture receiver (raw JPEG body + X-* headers). */
class MacClient(private val endpoint: MacEndpoint) {
    private val client = OkHttpClient.Builder()
        .connectTimeout(5, TimeUnit.SECONDS)
        .callTimeout(30, TimeUnit.SECONDS)
        .build()
    // Short-timeout client for the pre-pairing reachability probe only (fast, honest failure); the
    // 30s upload timeout above is untouched so large photo POSTs still have time.
    private val preflight = OkHttpClient.Builder()
        .connectTimeout(3, TimeUnit.SECONDS)
        .callTimeout(4, TimeUnit.SECONDS)
        .build()

    private fun auth() = "Bearer ${endpoint.token}"

    fun ping(): Boolean = try {
        val req = Request.Builder().url("${endpoint.baseUrl}/ping")
            .header("Authorization", auth()).get().build()
        client.newCall(req).execute().use { it.isSuccessful }
    } catch (e: Exception) {
        false
    }

    /** Classify reachability of the Mac for a clear pairing diagnostic (short timeout). */
    fun reachability(): Reachability = try {
        val req = Request.Builder().url("${endpoint.baseUrl}/ping")
            .header("Authorization", auth()).get().build()
        preflight.newCall(req).execute().use { resp ->
            when {
                resp.isSuccessful -> Reachability.OK
                resp.code == 401 -> Reachability.UNAUTHORIZED
                else -> Reachability.REFUSED
            }
        }
    } catch (e: java.net.SocketTimeoutException) {
        Reachability.UNREACHABLE               // black-holed SYN — the AP-isolation signature
    } catch (e: java.net.ConnectException) {
        Reachability.REFUSED                   // RST — reachable host, nothing listening
    } catch (e: Exception) {
        Reachability.UNREACHABLE               // no route / unknown host → treat as unreachable
    }

    /** POST one JPEG (raw body) with grouping + minimal-tag headers. Returns true on 2xx. */
    fun postPhoto(
        jpeg: ByteArray, group: String, seq: Int, type: String,
        priority: String?, year: Int?, month: Int?, device: String,
        replaces: String? = null
    ): Boolean = try {
        val body = jpeg.toRequestBody("image/jpeg".toMediaType())
        val b = Request.Builder().url("${endpoint.baseUrl}/photo")
            .header("Authorization", auth())
            .header("X-Group", group)
            .header("X-Seq", seq.toString())
            .header("X-Type", type)
            .header("X-Device", device)
        if (!priority.isNullOrBlank()) b.header("X-Priority", priority)
        if (year != null) b.header("X-Year", year.toString())
        if (month != null) b.header("X-Month", month.toString())
        // The old group this photo replaces (reclassify) — the Mac drops the orphaned old copy.
        if (!replaces.isNullOrBlank()) b.header("X-Replaces", replaces)
        client.newCall(b.post(body).build()).execute().use { it.isSuccessful }
    } catch (e: Exception) {
        false
    }

    fun sessionComplete(): Boolean = try {
        val req = Request.Builder().url("${endpoint.baseUrl}/session/complete")
            .header("Authorization", auth())
            .post(ByteArray(0).toRequestBody(null)).build()
        client.newCall(req).execute().use { it.isSuccessful }
    } catch (e: Exception) {
        false
    }

    /** End of a document segment: pages already streamed in via postPhoto; this tells the Mac the group
     *  is complete (so its tag card can appear) and carries the segment's tags. No image bytes. */
    fun segmentComplete(group: String, priority: String?, year: Int?, month: Int?): Boolean = try {
        val b = Request.Builder().url("${endpoint.baseUrl}/segment/complete")
            .header("Authorization", auth())
            .header("X-Group", group)
        if (!priority.isNullOrBlank()) b.header("X-Priority", priority)
        if (year != null) b.header("X-Year", year.toString())
        if (month != null) b.header("X-Month", month.toString())
        client.newCall(b.post(ByteArray(0).toRequestBody(null)).build()).execute().use { it.isSuccessful }
    } catch (e: Exception) {
        false
    }

    /** Best-effort notice that the phone is re-pairing, so the Mac re-shows the pairing QR instead of
     *  sitting on a stale "paired" state. Fire-and-forget (may not reach the Mac if the link is already down). */
    fun sessionDisconnect(): Boolean = try {
        val req = Request.Builder().url("${endpoint.baseUrl}/session/disconnect")
            .header("Authorization", auth())
            .post(ByteArray(0).toRequestBody(null)).build()
        client.newCall(req).execute().use { it.isSuccessful }
    } catch (e: Exception) {
        false
    }
}

File: ArchiveCaptureiOS/Sources/ArchiveCaptureiOS/Net/MacClient.swift
import Foundation

/// Why a pairing attempt failed — so the connect UI can name the cause + the fix instead of a bare
/// "couldn't connect". `unreachable` = timeout / no route (the client-isolation case, and on iOS also
/// a refused connection, which surfaces as cannotConnectToHost); `refused` = reached a server but its
/// reply wasn't our happy path; `unauthorized` = reached the Mac but the token was rejected (HTTP 401).
/// Mirrors the Android `Reachability` enum + the plan's `ConnectPhase` causes.
enum ConnectResult { case ok, unauthorized, refused, unreachable }

/// Protocol-v2 client for the Archive Processor Live Capture receiver (raw JPEG body + X-* headers).
/// Mirrors the Android `MacClient` and the Mac's `CaptureServer` routes exactly.
struct MacClient {
    let endpoint: MacEndpoint

    /// `timeout` is parametrized so the pre-pairing reachability probe can fail fast (~3.5s) while photo
    /// uploads keep the full 30s — never shorten the upload timeout to speed up the probe.
    private func makeRequest(_ path: String, method: String, timeout: TimeInterval = 30) -> URLRequest? {
        guard let url = URL(string: "\(endpoint.baseURL)\(path)") else { return nil }
        var req = URLRequest(url: url, timeoutInterval: timeout)
        req.httpMethod = method
        req.setValue("Bearer \(endpoint.token)", forHTTPHeaderField: "Authorization")
        return req
    }

    private static func isSuccess(_ resp: URLResponse?) -> Bool {
        (resp as? HTTPURLResponse).map { (200..<300).contains($0.statusCode) } ?? false
    }

    func ping() async -> Bool {
        guard let req = makeRequest("/ping", method: "GET") else { return false }
        do { let (_, resp) = try await URLSession.shared.data(for: req); return Self.isSuccess(resp) }
        catch { return false }
    }

    /// Classify reachability of the Mac for a clear pairing diagnostic. Uses a short timeout so a
    /// black-holed SYN (client-isolation) fails fast; the 30s upload timeout is deliberately untouched.
    func reachability(timeout: TimeInterval = 3.5) async -> ConnectResult {
        guard let req = makeRequest("/ping", method: "GET", timeout: timeout) else { return .unreachable }
        do {
            let (_, resp) = try await URLSession.shared.data(for: req)
            guard let http = resp as? HTTPURLResponse else { return .refused }
            if (200..<300).contains(http.statusCode) { return .ok }
            if http.statusCode == 401 { return .unauthorized }
            return .refused
        } catch {
            // A thrown URLError (timedOut / cannotConnectToHost / networkConnectionLost / no route) means
            // we never got an HTTP reply — the client-isolation / no-route / server-not-started signature.
            // iOS can't reliably split "refused" from "unreachable" (a refused TCP connection also surfaces
            // as cannotConnectToHost), so we report unreachable, whose guidance also covers "start Live
            // Capture". `refused` is reserved for a reached-but-non-happy HTTP response.
            return .unreachable
        }
    }

    /// POST one JPEG (raw body) with grouping + minimal-tag headers. Returns true on 2xx.
    /// The Mac requires a non-empty X-Group and a numeric X-Seq (else 400), so both are always sent.
    func postPhoto(jpeg: Data, group: String, seq: Int, type: String,
                   priority: String?, year: Int?, month: Int?, device: String,
                   replaces: String? = nil) async -> Bool {
        guard var req = makeRequest("/photo", method: "POST") else { return false }
        req.setValue("image/jpeg", forHTTPHeaderField: "Content-Type")
        req.setValue(group, forHTTPHeaderField: "X-Group")
        req.setValue(String(seq), forHTTPHeaderField: "X-Seq")
        req.setValue(type, forHTTPHeaderField: "X-Type")
        req.setValue(device, forHTTPHeaderField: "X-Device")
        // The old group this photo replaces (reclassify) — the Mac drops the orphaned old copy.
        if let replaces, !replaces.isEmpty { req.setValue(replaces, forHTTPHeaderField: "X-Replaces") }
        if let p = priority, !p.trimmingCharacters(in: .whitespaces).isEmpty {
            req.setValue(p, forHTTPHeaderField: "X-Priority")
        }
        if let y = year { req.setValue(String(y), forHTTPHeaderField: "X-Year") }
        if let m = month { req.setValue(String(m), forHTTPHeaderField: "X-Month") }
        do { let (_, resp) = try await URLSession.shared.upload(for: req, from: jpeg); return Self.isSuccess(resp) }
        catch { return false }
    }

    func sessionComplete() async -> Bool {
        guard let req = makeRequest("/session/complete", method: "POST") else { return false }
        do { let (_, resp) = try await URLSession.shared.upload(for: req, from: Data()); return Self.isSuccess(resp) }
        catch { return false }
    }

    /// End of a document segment: its pages already streamed in via `postPhoto`; this tells the Mac the
    /// group is complete (so its tag card can appear) and carries the segment's tags. No image bytes.
    func segmentComplete(group: String, priority: String?, year: Int?, month: Int?) async -> Bool {
        guard var req = makeRequest("/segment/complete", method: "POST") else { return false }
        req.setValue(group, forHTTPHeaderField: "X-Group")
        if let p = priority, !p.trimmingCharacters(in: .whitespaces).isEmpty {
            req.setValue(p, forHTTPHeaderField: "X-Priority")
        }
        if let y = year { req.setValue(String(y), forHTTPHeaderField: "X-Year") }
        if let m = month { req.setValue(String(m), forHTTPHeaderField: "X-Month") }
        do { let (_, resp) = try await URLSession.shared.upload(for: req, from: Data()); return Self.isSuccess(resp) }
        catch { return false }
    }

    /// Best-effort notice that the phone is re-pairing, so the Mac re-shows the pairing QR instead of
    /// sitting on a stale "paired" state. Fire-and-forget (may not reach the Mac if the link is already down).
    func sessionDisconnect() async -> Bool {
        guard let req = makeRequest("/session/disconnect", method: "POST") else { return false }
        do { let (_, resp) = try await URLSession.shared.upload(for: req, from: Data()); return Self.isSuccess(resp) }
        catch { return false }
    }
}
